### ISSUE TITLE
Enable TypeScript in NodeJS code

### ISSUE THREAD

**Issue by samreid** [phetsims/chipper#1437]
I swamped https://github.com/phetsims/chipper/issues/1272 with commits, so let's continue over here. In summary, we would like to enable TypeScript usage in perennial & chipper. Recent dev meeting notes are in: https://docs.google.com/document/d/11Gt3Ulalj0fCD2fFeCjPT5ni_9mM2WjkGc4ysisQmo8/edit

A summary is in https://github.com/phetsims/chipper/issues/1272#issuecomment-2050288975

> * @zepumph is doing regular work in perennial and chipper and feels like it would be a significant productivity improvement to have the code in TS.
> * @samreid feels that he has taken it about as far as he can on his own and would like help from either @jonathanolson, @pixelzoom, or @zepumph.
> 
> The general conclusion is that we should definitely go for an incremental process, and @samreid will review everyone's input in the developer notes doc and we will subsequently discuss the incremental plan, and @marlitas will get this item on the planning list for the next iteration.  (UPDATE: or the current iteration)

From https://github.com/phetsims/chipper/issues/1272#issuecomment-2050453575

> Also, the revelation that we can use import statements anywhere in the chain and everything will still work because the transpiler outputs require statements that grunt can consume.

Current checklist from https://github.com/phetsims/chipper/issues/1272#issuecomment-2051005300

Next steps:

- [x] Apply the patch above
- [x] Address some of the TODOs, such as standardizing the status cache keys
- [x] Notify the dev team that changes are incoming
- [x] Commit and push to main
- [x] Test on Windows
- [x] Notify team members that the transpiler is now outputting a commonjs/requirejs format for some repos. Notify that cache uses different keys and will be invalid.  Recommend restarting transpiler with --clean.  Notify to be on the lookout for gruntMain issues.
- [x] Point Gruntfile entry points at gruntMain
- [x] test a chains build/deploy process
- [x] test on Windows again
- [ ] adjust tsconfigs to optimize for usage in node
- [ ] port 1-2 grunt files to typescript minimally
- [ ] move gruntMain to perennial-alias
- [ ] move perennial to use perennial/gruntMain instead of js/grunt/Gruntfile.js
- [ ] Support other non chipper gruntfiles (aqua/rosetta/etc)
- [ ] Support gruntfiles that use chipper AND register their own tasks (dot/kite/scenery)
- [ ] port 1-2 non-grunt files minimally, and consider how we should launch it. (ts-node/deno/custom)
- [ ] How do we determine what the list of build-tool dependencies are to transpile? aqua needs perennial, chipper needs perennial-alias?
- [ ] toggle gruntMain to use typescript
- [ ] sourcemaps? how will stack traces look?
- [ ] more complete porting of grunt and non-grunt files in chipper
- [ ] porting some files in perennial and perennial alias
- [ ] what if we want to load Property.ts in chipper? What would that take?
- [ ] For winston, we could use a babel plugin or output to repo/dist
- [ ] Several TODOs in the code (9 on 5/15/24)
- [ ] Can we use tsc for our transpiling yet? Probably not, and likely it shouldn't be part of this build-tool work, but it may be worth investigating. Babel seems good and fast enough for our purposes now.

**Comment by samreid** [comment]
@zepumph and I reviewed the current status and agreed it is at a good point to put on hold until we have more bandwidth to "flip the switch".

**Comment by zepumph** [comment]
We want to prioritize https://github.com/phetsims/chipper/issues/1356 first.

**Comment by zepumph** [comment]
@samreid and I met today to discuss this further. We had the following discussions:
1. ts-node may be worth investigating to drive our node-based typescript, but we were totally thwarted even by a hello world, and see no path forward to getting it to work with grunt.
2. Transpiling with our custom transpiler into two modes seems good and reasonable as a step forward. We don't need to investigate other strategies to unblock this issue (but could at a different time independently). TSC will most likely never be as fast as the babel script we are using in Transpiler.js.

**Comment by zepumph** [comment]
What will it take to commit flipping the feature flag in gruntMain?

- [x] confirm that grunt options work
- [x] confirm that tasks registered by a "sub gruntfile" work (like aqua's `quick-server` task)
- [ ] `cd buoyancy; grunt/` gave this problem: 

```
 mjkauzmann ~/PHET/git/buoyancy (main)
 $ grunt --lint=false
Running "report-media" task

Running "clean" task

Running "build" task
Fatal error: Perennial task failed:
Error: ENOENT: no such file or directory, open 'C:\Users\mjkauzmann\PHET\git\chipper\dist\commonjs\perennial-alias\data\active-repos'
    at Object.openSync (node:fs:601:3)
    at Object.readFileSync (node:fs:469:35)
    at Object.<anonymous> (C:\Users\mjkauzmann\PHET\git\chipper\dist\commonjs\chipper\js\grunt\webpackBuild.js:29:22)
    at Module._compile (node:internal/modules/cjs/loader:1254:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1308:10)
    at Module.load (node:internal/modules/cjs/loader:1117:32)
    at Module._load (node:internal/modules/cjs/loader:958:12)
    at Module.require (node:internal/modules/cjs/loader:1141:19)
    at require (node:internal/modules/cjs/helpers:110:18)
    at Object.<anonymous> (C:\Users\mjkauzmann\PHET\git\chipper\dist\commonjs\chipper\js\grunt\buildStandalone.js:26:20)
Full Error details:
Error: ENOENT: no such file or directory, open 'C:\Users\mjkauzmann\PHET\git\chipper\dist\commonjs\perennial-alias\data\active-repos'
```

**Comment by samreid** [comment]
It seems like data/active-repos is accessed through a relative path. Two ideas:

* Should we copy perennial/data files to chipper/dist?
* Should we change the code so it can find active-repos and other data files no matter where things are run from?

**Comment by samreid** [comment]
I added some commits in a branch. After the change above, `grunt lint` succeeds in buoyancy.

**Comment by samreid** [comment]
Node 22.6 was released today with support for --experimental-strip-types https://github.com/nodejs/node/blob/main/doc/changelogs/CHANGELOG_V22.md#22.6.0

**Comment by jessegreenberg** [comment]
8/15/24 - 

@samreid reviewed the above with the dev team so we are all aware and we can remember that option when it is time to work on TS support in chipper/perennial in a future iteration.

**Comment by samreid** [comment]
tsx worked well. For instance, adding tsx in chipper/package.json, then npm install, then rename Gruntfile.js => Gruntfile.ts and change the top line of `which grunt` (for me it is /Users/samreid/.npm-global/bin/grunt) from `#!/usr/bin/env node` to `#!/usr/bin/env npx tsx`

**Comment by samreid** [comment]
I cannot find a robust or maintainable way to "patch in" tsx or npx tsx or even --experimental-strip-types. So I'll continue with the manually transpiled code. Patch for testing:

<details>

```diff

Subject: [PATCH] Add migration rules for selfVisible for inadvertently omitted masses, sort each within screen, see https://github.com/phetsims/density-buoyancy-common/issues/388
---
Index: perennial/js/common/phetTimingLog.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/perennial/js/common/phetTimingLog.js b/perennial/js/common/phetTimingLog.js
--- a/perennial/js/common/phetTimingLog.js	(revision 420fddfb5383823533cda9b52427d4dc09d2f6bf)
+++ b/perennial/js/common/phetTimingLog.js	(date 1725632825116)
@@ -30,7 +30,7 @@
 const fs = require( 'fs' );
 const assert = require( 'assert' );
 
-const logDir = path.resolve( __dirname, '../../logs' );
+const logDir = path.resolve( process.cwd(), '../perennial-alias/logs' );
 try {
   fs.mkdirSync( logDir );
 }
Index: chipper/js/scripts/generateDevelopmentStrings.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/chipper/js/scripts/generateDevelopmentStrings.js b/chipper/js/scripts/generateDevelopmentStrings.js
--- a/chipper/js/scripts/generateDevelopmentStrings.js	(revision f4265c0399de95bc88b9e20e3f5e84baa9c8555f)
+++ b/chipper/js/scripts/generateDevelopmentStrings.js	(date 1725631831207)
@@ -20,7 +20,7 @@
 
   const start = Date.now();
 
-  const rootPath = path.join( __dirname, '..', '..', '..' );
+  const rootPath = path.join( process.cwd(), '..' );
 
   // OS-independent path to babel repo.
   const babelPath = path.join( rootPath, 'babel' );
Index: chipper/js/scripts/absolute-tsc.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/chipper/js/scripts/absolute-tsc.js b/chipper/js/scripts/absolute-tsc.js
--- a/chipper/js/scripts/absolute-tsc.js	(revision f4265c0399de95bc88b9e20e3f5e84baa9c8555f)
+++ b/chipper/js/scripts/absolute-tsc.js	(date 1725631831184)
@@ -49,7 +49,7 @@
 
   // console.log( 'changes detected...' );
 
-  const results = await execute( 'node', [ `${__dirname}/../../../chipper/node_modules/typescript/bin/tsc` ], args[ 0 ], {
+  const results = await execute( 'node', [ `${process.cwd()}/../chipper/node_modules/typescript/bin/tsc` ], args[ 0 ], {
     errors: 'resolve',
 
     // TODO: it would be nice not to need this, https://github.com/phetsims/chipper/issues/1415
Index: chipper/js/grunt/copySupplementalPhetioFiles.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/chipper/js/grunt/copySupplementalPhetioFiles.js b/chipper/js/grunt/copySupplementalPhetioFiles.js
--- a/chipper/js/grunt/copySupplementalPhetioFiles.js	(revision f4265c0399de95bc88b9e20e3f5e84baa9c8555f)
+++ b/chipper/js/grunt/copySupplementalPhetioFiles.js	(date 1725631831195)
@@ -709,7 +709,7 @@
     else {
 
       // output dir must be an absolute path
-      const outputDir = path.resolve( __dirname, `../../${repo}/${buildDir}` );
+      const outputDir = path.resolve( process.cwd(), `../${repo}/${buildDir}` );
 
       const compiler = webpack( {
         module: {
Index: chipper/js/grunt/webpackDevServer.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/chipper/js/grunt/webpackDevServer.js b/chipper/js/grunt/webpackDevServer.js
--- a/chipper/js/grunt/webpackDevServer.js	(revision f4265c0399de95bc88b9e20e3f5e84baa9c8555f)
+++ b/chipper/js/grunt/webpackDevServer.js	(date 1725631831201)
@@ -37,7 +37,7 @@
       const packageObject = JSON.parse( fs.readFileSync( packageFile, 'utf-8' ) );
       if ( packageObject.phet && packageObject.phet.requirejsNamespace ) {
         reposByNamespace[ packageObject.phet.requirejsNamespace ] = repo;
-        aliases[ packageObject.phet.requirejsNamespace ] = path.resolve( __dirname, `../../../${repo}${repo === 'brand' ? '/phet' : ''}/js` );
+        aliases[ packageObject.phet.requirejsNamespace ] = path.resolve( process.cwd(), `../${repo}${repo === 'brand' ? '/phet' : ''}/js` );
       }
     }
   }
@@ -65,7 +65,7 @@
     entry: entries,
 
     output: {
-      path: path.resolve( __dirname, `../../${ChipperConstants.BUILD_DIR}` ),
+      path: path.resolve( process.cwd(), `../chipper/${ChipperConstants.BUILD_DIR}` ),
       filename: '[name].js',
       publicPath: '/dist/'
     },
@@ -112,7 +112,7 @@
 
   const server = new WebpackDevServer( {
     static: {
-      directory: path.join( __dirname, '../../../' )
+      directory: path.join( process.cwd(), '../' )
     },
     compress: true,
     port: port,
Index: chipper/js/grunt/webpackBuild.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/chipper/js/grunt/webpackBuild.js b/chipper/js/grunt/webpackBuild.js
--- a/chipper/js/grunt/webpackBuild.js	(revision f4265c0399de95bc88b9e20e3f5e84baa9c8555f)
+++ b/chipper/js/grunt/webpackBuild.js	(date 1725631621986)
@@ -17,17 +17,17 @@
 // eslint-disable-next-line require-statement-match
 const { ModifySourcePlugin, ConcatOperation } = require( 'modify-source-webpack-plugin' );
 
-const activeRepos = fs.readFileSync( path.resolve( __dirname, '../../../perennial-alias/data/active-repos' ), 'utf-8' ).trim().split( /\r?\n/ ).map( s => s.trim() );
+const activeRepos = fs.readFileSync( path.resolve( process.cwd(), '../perennial-alias/data/active-repos' ), 'utf-8' ).trim().split( /\r?\n/ ).map( s => s.trim() );
 const reposByNamespace = {};
 const aliases = {};
 
 for ( const repo of activeRepos ) {
-  const packageFile = path.resolve( __dirname, `../../../${repo}/package.json` );
+  const packageFile = path.resolve( process.cwd(), `../${repo}/package.json` );
   if ( fs.existsSync( packageFile ) ) {
     const packageObject = JSON.parse( fs.readFileSync( packageFile, 'utf-8' ) );
     if ( packageObject.phet && packageObject.phet.requirejsNamespace ) {
       reposByNamespace[ packageObject.phet.requirejsNamespace ] = repo;
-      aliases[ packageObject.phet.requirejsNamespace ] = path.resolve( __dirname, `../../../${repo}${repo === 'brand' ? '/phet' : ''}/js` );
+      aliases[ packageObject.phet.requirejsNamespace ] = path.resolve( process.cwd(), `../${repo}${repo === 'brand' ? '/phet' : ''}/js` );
     }
   }
 }
@@ -52,7 +52,7 @@
  * @returns {Array.<string>}
  */
 const getRelativeModules = modules => {
-  const root = path.resolve( __dirname, '../../../' );
+  const root = path.resolve( process.cwd(), '../' );
   return modules
 
     // Webpack 5 reports intermediate paths which need to be filtered out
@@ -82,7 +82,7 @@
       outputDir: repo
     }, options );
 
-    const outputDir = path.resolve( __dirname, `../../${ChipperConstants.BUILD_DIR}`, options.outputDir );
+    const outputDir = path.resolve( process.cwd(), `../chipper/${ChipperConstants.BUILD_DIR}`, options.outputDir );
     const outputFileName = `${repo}.js`;
     const outputPath = path.resolve( outputDir, outputFileName );
 
Index: chipper/js/grunt/gruntMain.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/chipper/js/grunt/gruntMain.js b/chipper/js/grunt/gruntMain.js
--- a/chipper/js/grunt/gruntMain.js	(revision f4265c0399de95bc88b9e20e3f5e84baa9c8555f)
+++ b/chipper/js/grunt/gruntMain.js	(date 1725632089808)
@@ -14,7 +14,7 @@
 /* eslint-env node */
 
 // Switch between TypeScript and JavaScript (feature flag)
-const LAUNCH_FROM_CHIPPER_DIST = false;
+const LAUNCH_FROM_CHIPPER_DIST = true;
 
 if ( LAUNCH_FROM_CHIPPER_DIST ) {
 
Index: chipper/js/grunt/lint.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/chipper/js/grunt/lint.js b/chipper/js/grunt/lint.js
--- a/chipper/js/grunt/lint.js	(revision f4265c0399de95bc88b9e20e3f5e84baa9c8555f)
+++ b/chipper/js/grunt/lint.js	(date 1725631831177)
@@ -200,7 +200,7 @@
   }
 };
 
-const repoRootPath = path.join( __dirname, '../../../' ); // Will end in a slash
+const repoRootPath = path.join( process.cwd(), '../' ); // Will end in a slash
 const escaped = repoRootPath.replace( /\\/g, '\\\\' ); // Handle any backslashes in the path
 
 // Regex that captures the repo via the path

**Comment by samreid** [comment]
After the changes above, here is a patch that exercises typescript in the webpack build:

<details>

```diff

Subject: [PATCH] Use process.cwd() instead of __dirname, see https://github.com/phetsims/chipper/issues/1456
---
Index: js/grunt/webpackBuild.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/js/grunt/webpackBuild.js b/js/grunt/webpackBuild.ts
rename from js/grunt/webpackBuild.js
rename to js/grunt/webpackBuild.ts
--- a/js/grunt/webpackBuild.js	(revision 6519fe7ef018c99196b27b229bfdb7f1c7f54163)
+++ b/js/grunt/webpackBuild.ts	(date 1725635635862)
@@ -76,6 +76,9 @@
  * @returns {Promise.<string>} - The combined JS output from the process
  */
 const webpackBuild = function webpackBuild( repo, brand, options ) {
+
+  const number:string = 'seven';
+  console.log(number);
   return new Promise( ( resolve, reject ) => {
 
     options = _.merge( {
Index: js/grunt/gruntMain.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/js/grunt/gruntMain.js b/js/grunt/gruntMain.js
--- a/js/grunt/gruntMain.js	(revision 6519fe7ef018c99196b27b229bfdb7f1c7f54163)
+++ b/js/grunt/gruntMain.js	(date 1725635496670)
@@ -14,7 +14,7 @@
 /* eslint-env node */
 
 // Switch between TypeScript and JavaScript (feature flag)
-const LAUNCH_FROM_CHIPPER_DIST = false;
+const LAUNCH_FROM_CHIPPER_DIST = true;
 
 if ( LAUNCH_FROM_CHIPPER_DIST ) {
 
```

</details>

Perhaps we could start porting to typescript in a branch.

**Comment by samreid** [comment]
I tested on Windows and confirmed that `grunt --brands=phet-io` is working well with the transpiled chipper grunt code.

**Comment by samreid** [comment]
https://github.com/phetsims/chipper/issues/1459 is now the most promising way forward. Let's put this issue on hold until that issue is more fully explored. Note that we will likely want to revert some changes here, namely in the transpiler which is overcomplicated, once we are more confident in https://github.com/phetsims/chipper/issues/1459

**Comment by samreid** [comment]
After consultation with @zepumph, I feel the best way forward is the proposal in https://github.com/phetsims/chipper/issues/1459. I have reverted the various aspects for dual-mode transpiling for this issue, which were clumsy, slow, and overcomplicated. I notified slack dev-public to pull and restart transpiler watch processes. There are no more TODOs pointing to this issue. Let's close this issue and continue in #1459.

**Comment by samreid** [comment]
Well, let's keep this open (and don't delete the branch yet) since the branch has a lot of commits converting chipper/js/grunt/ files to typescript.

**Comment by samreid** [comment]
We recovered the commits and brought them into https://github.com/phetsims/chipper/issues/1459. Branch deleted. Closing.

**Comment by phet-dev** [comment]
Reopening because there is a TODO marked for this issue.

**Comment by samreid** [comment]
Fixed and will be merged to main with SWC, closing.
